monkey/monkey_island/cc/repositories/plugin_archive_parser.py

```diff
@@ -97,3 +138,55 @@ def _safe_extract_file(tar: TarFile, filename: str) -> IO[bytes]:
     # `Optional[IO[bytes]]`.
     assert file_obj is not None
     return file_obj
+
+
+def get_plugin_source_vendors(plugin_source_tar: TarFile) -> Sequence[TarInfo]:
+    plugin_source_vendors = [
+        member
+        for member in plugin_source_tar.getmembers()
+        if (member.name.startswith("vendor") and member.isdir())
+    ]
+    return plugin_source_vendors
+
+
+def get_os_specific_plugin_source_archives(
+    plugin_source_tar, plugin_source_vendors, path
+) -> Mapping[OperatingSystem, bytes]:
+
+    os_specific_plugin_source_archives = {}
+
+    plugin_directory = path / f"plugin-{random_filename()}"
+    create_secure_directory(plugin_directory)
+
+    for vendor in plugin_source_vendors:
+        if "linux" in vendor.name:
+            vendor_os = OperatingSystem.LINUX
+        elif "windows" in vendor.name:
+            vendor_os = OperatingSystem.WINDOWS
+        else:
+            logger.info(f"Operating system of vendor directory ({vendor.name}) not recognised")
+            continue
+
+        os_specific_plugin_dir_path = plugin_directory / vendor_os.value
+        create_secure_directory(os_specific_plugin_dir_path)
+        os_specific_plugin_source_path = os_specific_plugin_dir_path / "source"
+        create_secure_directory(os_specific_plugin_source_path)
+
+        contents = [
+            tarinfo
+            for tarinfo in plugin_source_tar.getmembers()
+            if tarinfo.name.startswith(f"{vendor.name}/") or not tarinfo.name.startswith("vendor")
+        ]
+        plugin_source_tar.extractall(members=contents, path=os_specific_plugin_source_path)
```

Comment:
We should be using `_safe_extract_file` or `_safe_extract_dir` instead of `extractall`